crates/rsonpath-lib/src/query/automaton.rs

```diff
@@ -18,8 +18,89 @@ pub struct Automaton<'q> {
     states: Vec<StateTable<'q>>,
 }
 
+/// Represent the distinct methods of moving on a match between states.
+#[derive(Debug, Copy, PartialEq, Clone, Eq)]
+pub enum TransitionLabel<'q> {
+    /// Wraps a textual field [`Label`] in a JSON object.
+    ObjectMember(&'q Label),
+    /// Wraps an array index [`NonNegativeArrayIndex`] in a JSON object.
+    ArrayIndex(NonNegativeArrayIndex),
+}
+
+impl<'q> TransitionLabel<'q> {
+    /// Return a display object with a UTF8 representation of this label.
+    #[must_use]
+    #[inline(always)]
+    pub fn display(&self) -> impl Display + 'q {
+        match self {
+            // This isn't just calling label.display() because the compile makes opaque types unique at the declaration site.
+            // So, the one for display() that is defined above would not equate to Label's def.
+            TransitionLabel::ObjectMember(label) => format!("{}", label.display()),
+            TransitionLabel::ArrayIndex(index) => format!("{index}"),
+        }
+    }
+
+    ///Return the textual [`Label`] being wrapped if so.  Returns [`None`] otherwise.
+    #[must_use]
+    #[inline(always)]
+    pub fn get_label(&self) -> Option<&Label> {
+        match self {
+            TransitionLabel::ObjectMember(l) => Some(l),
+            TransitionLabel::ArrayIndex(_) => None,
+        }
+    }
+
+    /// Consumes the [`TransitionLabel`] and gives the wrapped [`Label`], if so.  Returns [`None`] otherwise.
+    #[must_use]
+    #[inline(always)]
+    pub fn get_label_owned(self) -> Option<&'q Label> {
+        match self {
+            TransitionLabel::ObjectMember(l) => Some(l),
+            TransitionLabel::ArrayIndex(_) => None,
+        }
+    }
+
+    /// Wraps a [`Label`] in a [`TransitionLabel`].
+    #[must_use]
+    #[inline(always)]
+    pub fn new_object_member(label: &'q Label) -> Self {
+        TransitionLabel::ObjectMember(label)
+    }
+
+    /// Wraps a [`NonNegativeArrayIndex`] in a [`TransitionLabel`].
+    #[must_use]
+    #[inline(always)]
+    pub fn new_array_index(label: NonNegativeArrayIndex) -> Self {
+        TransitionLabel::ArrayIndex(label)
+    }
+}
+
+impl<'q> From<&'q Label> for TransitionLabel<'q> {
+    #[must_use]
+    #[inline(always)]
+    fn from(label: &'q Label) -> Self {
+        TransitionLabel::new_object_member(label)
+    }
+}
+
+impl From<&NonNegativeArrayIndex> for TransitionLabel<'_> {
+    #[must_use]
+    #[inline(always)]
+    fn from(label: &NonNegativeArrayIndex) -> Self {
+        TransitionLabel::new_array_index(*label)
+    }
+}
+
+impl From<NonNegativeArrayIndex> for TransitionLabel<'_> {
+    #[must_use]
+    #[inline(always)]
+    fn from(label: NonNegativeArrayIndex) -> Self {
+        TransitionLabel::new_array_index(label)
+    }
+}
```

Comment:
Rust ProTip™: if you implement a trait for `Borrow<T>` you'll get an implementation that accepts `T`, `&T`, `Box<T>`, `Arc<T>`, and most other "wrapper" types for free!

Reply:
I've added the appropriate `From` implementation.